functional/src/Logic.hs

```diff
@@ -12,29 +13,26 @@ switchPlayer game =
       PlayerX -> game { gamePlayer = PlayerO }
       PlayerO -> game { gamePlayer = PlayerX }
 
-playerWon :: Player -> Board -> Bool
-playerWon player board = any isVictoryProj projs
-    where projs = allRowCoords
-                  ++ allColumnCoords
-                  ++ allDiagCoords
-          allRowCoords = [[(i, j) | j <- [0 .. n - 1]] | i <- [0 .. n - 1]]
-          allColumnCoords = [[(j, i) | j <- [0 .. n - 1]] | i <- [0 .. n - 1]]
-          allDiagCoords = [ [(i, i) | i <- [0 .. n - 1]]
-                          , [(i, n - 1 - i) | i <- [0 .. n - 1]]
-                          ]
-          isVictoryProj proj = (n ==)
-                               $ length
-                               $ filter (\cell -> cell == Full player)
-                               $ map (\coord -> board ! coord) proj
+full :: [Cell] -> Maybe Player
+full (x@(Just player):xs) | and $ map (== x) xs = Just player
```

Comment:
Interesting point! Your version is simpler, I like that. I think it does hide the intent a bit thought.

The current version tells the story of `full` returning either `Nothing` or `Just player`. You don't have to look into `Cell` to understand it, it is self-contained. The proposed change returns whatever is there, as long as the sequence is made of copies of it. To know what you can get out of Full you now have to look into the other file.

Another minor point that you didn't mention which probably means you already agree on: I believe that GHC is sufficiently smart not to unpack the cell and create a new copy to return. So I kept the `= Just Player` that aligns better with `= Nothing` than `= x` would.

Good catch on `all` 👍